### PR TITLE
Enable building on Linux

### PR DESCRIPTION
**Pull request by omajid**
This commit adds a new build.sh file (based on the Build.cmd file) to allow building roslyn-analzysers on Linux using .NET 5.

This repository likes to use `$(MSBuildProjectFullPath)` as the PackageId.  This has a problem on Linux because the file path separator (`/`) is also used the version separator when the package id is placed in the assets.json file. This breaks the build with an error like

    roslyn-analyzers/.dotnet/sdk/5.0.100-rc.2.20479.15/Sdks/Microsoft.NET.Sdk/targets/Microsoft.PackageDependencyResolution.targets(241,5): error NETSDK1060: Error reading assets file: Error loading lock file 'roslyn-analyzers/artifacts/obj/Microsoft.CodeQuality.CSharp.Analyzers/project.assets.json': 'roslyn-analyzers/src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/Microsoft.CodeQuality.Analyzers.csproj*/3.3.2-dev' is not a valid version string. (Parameter 'value') [roslyn-analyzers/src/NetAnalyzers/CSharp/Microsoft.CodeQuality.Analyzers/Microsoft.CodeQuality.CSharp.Analyzers.csproj]

Fix that by converting `/` to `\` on Linux.

The build also complains about a bunch of missing AnalyzerReleases.Shipped.md and AnalyzerReleases.Unshipped.md files, so I added them.

### CODE REVIEW

**Comment by omajid** [comment]
This PR does not enable `./build.sh --pack` to work. That requires additional changes because `csi.exe` is not available in the .NET Core or .NET 5 SDK. 

@dagood suggested I port `./eng/GenerateAnalyzerNuspec.csx` to an msbuild task to work around that.

This is a (cleaned-up) subset of a fairly hacky PR in source-build: https://github.com/dotnet/source-build/pull/1849

cc @dagood @crummel @dseefeld @tmds

**Comment by omajid** [comment]
cc @jmarolf as the source-build champion for roslyn-analyzers.

**Comment by jmarolf** [comment]
@mavasani 

> The build also complains about a bunch of missing AnalyzerReleases.Shipped.md and AnalyzerReleases.Unshipped.md files, so I added them.

Is this the right thing to do?

**Comment by mavasani** [comment]
> The build also complains about a bunch of missing AnalyzerReleases.Shipped.md and AnalyzerReleases.Unshipped.md files, so I added them.

No, it is not. Let me look at the PR.

**Comment by mavasani** [comment]
@omajid So, I looked at the added AnalyzerReleases files and it seems all of them were added for Setup/VSIX projects. These projects do not and will not contain any analyzers, so they do not need analyzer release tracking. Following check ensures release tracking is skipped for Setup/VSIX projects: https://github.com/dotnet/roslyn-analyzers/blob/2ac1ca7c1c8fb0f4ee2623ba6c903686681ce44f/Directory.Build.targets#L21-L23

It seems like `IsVsixProject` is not being set of Setup projects in Linux? If so, another approach could be:

1. Revert the newly added AnalyzerReleases shipped and unshipped files
2. Add `<ReleaseTrackingOptOut>true</ReleaseTrackingOptOut>` to each of the Setup projects in the repo. See https://github.com/dotnet/roslyn-analyzers/search?q=%3CReleaseTrackingOptOut%3Etrue%3C%2FReleaseTrackingOptOut%3E

**Comment by omajid** [comment]
> It seems like IsVsixProject is not being set of Setup projects in Linux?

Yes, it seems like nothing is setting `IsVsixProject` on Linux.

```
$ cd src/Microsoft.CodeAnalysis.Analyzers/Setup
$ ../../../.dotnet/dotnet msbuild /pp: | grep -i vsixproject
    <IsSwixProject Condition="'$(VisualStudioInsertionComponent)' != '' and '$(IsVsixProject)' != 'true'">true</IsSwixProject>
  <!--<Import Project="VisualStudio.targets" Condition="'$(UsingToolVSSDK)' == 'true' and ('$(IsVsixProject)' == 'true' or '$(IsSwixProject)' == 'true' or '$(GeneratePkgDefFile)' == 'true') and '$(MSBuildRuntimeType)' != 'Core'" />-->
  <PropertyGroup Condition="'$(IsTestProject)' == 'true' or '$(NonShipping)' == 'true' or '$(IsVsixProject)' == 'true' or '$(FxCopAnalyzersProject)' == 'true'">
```

Even the string `IsVsixProject` only appears in the vstest assembly in the .NET SDK on Linux:

```
$  grep -r vsix ../../../.dotnet/
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/it/vstest.console.resources.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/ja/vstest.console.resources.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/ko/vstest.console.resources.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/fr/vstest.console.resources.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/cs/vstest.console.resources.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/tr/vstest.console.resources.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/zh-Hans/vstest.console.resources.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/pl/vstest.console.resources.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/vstest.console.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/es/vstest.console.resources.dll matches
Binary file ../../../.dotnet/sdk/5.0.100-rc.2.20479.15/pt-BR/vstest.console.resources.dll matches
```

**Comment by omajid** [comment]
CI seems to be stuck? It says Queued/Waiting bug also "Started 16h 3m 33s ago"..

File: nuget/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.Package.csproj
Comment (sharwell) [inline nuget/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.Package.csproj:23]: ❗ Just use the project file full name (including the extension). This is simpler, shorter, and avoids the need to explain platform-specific workarounds.

```xml
<PackageId>*$(MSBuildProjectFile)*</PackageId>
```

**Review by sharwell — CHANGES_REQUESTED** [review]

File: nuget/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.Package.csproj
Comment (dagood) [reply nuget/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.Package.csproj:23]: There's a comment describing why the repo did it this way--@sharwell why doesn't it apply (anymore)?

https://github.com/dotnet/roslyn-analyzers/blob/0919dbcf9bb03874aec10ec190c2f7f2397c10fb/nuget/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.Package.csproj#L13-L17

Comment (omajid) [reply nuget/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.Package.csproj:23]: I think keeping the file name (including `.csproj`) still keeps the comment valid?

Comment (dagood) [reply nuget/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.Package.csproj:23]: I see, I think I misunderstood "The package ID here doesn't matter" 😛. It obviously does matter because it lets us use this workaround, it just doesn't need to match anything else.

Comment (sharwell) [reply nuget/Microsoft.CodeAnalysis.Analyzers/Microsoft.CodeAnalysis.Analyzers.Package.csproj]: The package ID only matters in the sense that it meets both of the following two conditions:

1. It is not the same as the package ID for any other project in the solution.
2. It is not identical to the project file name _without_ extension. Even if we didn't use the file name with extension, the leading and trailing `*` characters would meet this condition.

File: build.sh
Comment (omajid) [inline build.sh:15]: The first 15 lines are taken from https://github.com/dotnet/arcade/blob/master/build.sh

**Comment by omajid** [comment]
Does the updated patch look okay?

**Comment by jmarolf** [comment]
I would like @sharwell and @mavasani to sign off before we merge this. We should also add a test leg that builds on linux at least

**Review by mavasani — APPROVED** [review]
LGTM.

Agree with @jmarolf that we should add a linux test leg to verify this does not regress.

**Comment by mavasani** [comment]
@sharwell Can you please review/unblock the PR?

**Comment by omajid** [comment]
> Agree with @jmarolf that we should add a linux test leg to verify this does not regress.

Would you like me to do this as part of this PR? Or as something separate?

**Comment by mavasani** [comment]
Separate PR would be fine. I believe @JoeRobich attempted the same here: https://github.com/dotnet/roslyn-analyzers/pull/2434

**Comment by omajid** [comment]
Can anyone please review and merge this PR? I have a followup that fixes `./build.sh --pack` too: https://github.com/omajid/roslyn-analyzers/commit/96b7a29b874554b9a3996efecc206906454eb275. I would like to get both reviewed and merged so we can use them in source-build.

**Comment by omajid** [comment]
Thanks, @mavasani !

**Pull request merged**
